### PR TITLE
Revise call and construct operation based on ES6

### PR DESCRIPTION
**Pull request by clover2123**
Signed-off-by: HyukWoo Park <hyukwoo.park@samsung.com>

### CODE REVIEW

**Comment by clover2123** [comment]
After this patch is landed, I'll update following features based on the ES6 spec.
- bound function
- builltin function

File: src/runtime/FunctionObject.cpp [hunk 1]
```diff
@@ -270,54 +270,49 @@ NEVER_INLINE void FunctionObject::generateBytecodeBlock(ExecutionState& state)
...
+        // FIXME: If kind is "base", then
+    } else {
+        // FIXME: Let thisArgument be OrdinaryCreateFromConstructor(newTarget, "%ObjectPrototype%").
+        thisArgument = new Object(state);
     }
 
-    Object* receiver;
-    if (cb->hasCallNativeFunctionCode()) {
-        receiver = cb->nativeFunctionData()->m_ctorFn(state, cb, argc, argv);
+    if (constructor->getFunctionPrototype(state).isObject()) {
```
Comment (yichoi) [inline src/runtime/FunctionObject.cpp]: For readability / State expression explicitly
if (constructor->getFunctionPrototype(state).isObject() == true)

Comment (clover2123) [reply src/runtime/FunctionObject.cpp]: Fixed it. Thanks.

File: src/runtime/FunctionObject.cpp [hunk 2]
```diff
@@ -270,54 +270,49 @@ NEVER_INLINE void FunctionObject::generateBytecodeBlock(ExecutionState& state)
...
-            return callee.asPointerValue()->asProxyObject()->call(state, callee, receiver, argc, argv);
-#endif
-    }
-    ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, errorMessage_Call_NotFunction);
-    return Value();
-}
-
-Object* FunctionObject::newInstance(ExecutionState& state, const size_t argc, Value* argv)
+// https://www.ecma-international.org/ecma-262/6.0/#sec-ecmascript-function-objects-construct-argumentslist-newtarget
+Object* FunctionObject::construct(ExecutionState& state, const size_t argc, Value* argv, const Value& newTarget)
```
Comment (yichoi) [inline src/runtime/FunctionObject.cpp]: Check pointer-type parameter / check before deferencing
ASSERT(argv != nullptr)

File: src/runtime/FunctionObject.cpp [hunk 3]
```diff
@@ -270,54 +270,49 @@ NEVER_INLINE void FunctionObject::generateBytecodeBlock(ExecutionState& state)
...
 
-    Value res = processCall(state, receiver, argc, argv, true);
-    if (res.isObject())
-        return res.asObject();
-    else
-        return receiver;
+    return thisArgument;
 }
 
 Value FunctionObject::processCall(ExecutionState& state, const Value& receiverSrc, const size_t argc, Value* argv, bool isNewExpression)
```
Comment (yichoi) [inline src/runtime/FunctionObject.cpp]: ditto

File: src/runtime/FunctionObject.h [hunk 4]
```diff
@@ -169,20 +169,41 @@ class FunctionObject : public Object {
         return FunctionKind::Normal;
     }
 
-    Value call(ExecutionState& state, const Value& receiver, const size_t argc, Value* argv)
+    // FIXME : switch to private member function
+    Value processCall(ExecutionState& state, const Value& receiver, const size_t argc, Value* argv, bool isNewExpression);
+
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-ecmascript-function-objects-call-thisargument-argumentslist
+    virtual Value call(ExecutionState& state, const Value& thisValue, const size_t argc, Value* argv) override
```
Comment (yichoi) [inline src/runtime/FunctionObject.h]: ditto

File: src/runtime/FunctionObject.h [hunk 5]
```diff
@@ -169,20 +169,41 @@ class FunctionObject : public Object {
...
+        return processCall(state, thisValue, argc, argv, false);
     }
-    // ECMAScript new operation
-    Object* newInstance(ExecutionState& state, const size_t argc, Value* argv);
 
-    ALWAYS_INLINE static Value call(ExecutionState& state, const Value& callee, const Value& receiver, const size_t argc, Value* argv, bool isNewExpression = false)
+    virtual Object* construct(ExecutionState& state, const size_t argc, Value* argv, const Value& newTarget) override;
+
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-call
+    ALWAYS_INLINE static Value call(ExecutionState& state, const Value& callee, const Value& thisValue, const size_t argc, Value* argv)
```
Comment (yichoi) [inline src/runtime/FunctionObject.h]: ditto

File: src/runtime/FunctionObject.h [hunk 6]
```diff
@@ -169,20 +169,41 @@ class FunctionObject : public Object {
...
+        // If IsCallable(F) is false, throw a TypeError exception.
+        if (callee.isCallable() == false) {
+            ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, errorMessage_NOT_Callable);
         }
+        // Return F.[[Call]](V, argumentsList).
+        return callee.asObject()->call(state, thisValue, argc, argv);
+    }
+
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-construct
+    ALWAYS_INLINE static Object* construct(ExecutionState& state, const Value& constructor, const size_t argc, Value* argv, Value newTarget = Value(Value::EmptyValue))
```
Comment (yichoi) [inline src/runtime/FunctionObject.h]: ditto

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 7]
```diff
@@ -115,7 +115,7 @@ static Value arraySpeciesCreate(ExecutionState& state, Object* originalArray, co
     }
     // Return Construct(C, <<length>>).
     Value argv[1] = { Value(length) };
-    return ByteCodeInterpreter::newOperation(state, C, 1, argv);
+    return FunctionObject::construct(state, C, 1, argv);
 }
 
 static Value builtinArrayIsArray(ExecutionState& state, Value thisValue, size_t argc, Value* argv, bool isNewExpression)
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: ditto

File: src/runtime/GlobalObjectBuiltinTypedArray.cpp [hunk 8]
```diff
@@ -904,7 +904,7 @@ static Value builtinTypedArraySubArray(ExecutionState& state, Value thisValue, s
     // Let argumentsList be «buffer, beginByteOffset, newLength».
     Value args[3] = { buffer, Value(beginByteOffset), Value(newLength) };
     // Return Construct(constructor, argumentsList).
-    return ByteCodeInterpreter::newOperation(state, constructor, 3, args);
+    return FunctionObject::construct(state, constructor, 3, args);
 }
 
 static Value builtinTypedArrayEvery(ExecutionState& state, Value thisValue, size_t argc, Value* argv, bool isNewExpression)
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinTypedArray.cpp]: ditto

File: src/interpreter/ByteCodeInterpreter.cpp [hunk 9]
```diff
@@ -1472,18 +1472,10 @@ NEVER_INLINE Value ByteCodeInterpreter::modOperation(ExecutionState& state, cons
 
 NEVER_INLINE Object* ByteCodeInterpreter::newOperation(ExecutionState& state, const Value& callee, size_t argc, Value* argv)
 {
```
Comment (yichoi) [inline src/interpreter/ByteCodeInterpreter.cpp]: Check pointer-type parameter / check before deferencing
ASSERT(argv != nullptr)

File: src/runtime/FunctionObject.cpp [hunk 10]
```diff
@@ -342,7 +337,7 @@ Value FunctionObject::processCall(ExecutionState& state, const Value& receiverSr
     }
 
     if (UNLIKELY(isSuperCall && isBuiltin() && !isNewExpression)) {
```
Comment (yichoi) [inline src/runtime/FunctionObject.cpp]: For readability / State expression explicitly
if (UNLIKELY(isSuperCall == true && isBuiltin() == true && isNewExpression == false))

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 11]
```diff
@@ -166,7 +166,7 @@ static Value builtinArrayFrom(ExecutionState& state, Value thisValue, size_t arg
         // If IsConstructor(C) is true, then
         if (C.isConstructor()) {
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: For readability / State expression explicitly
if (C.isConstructor() == true)

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 12]
```diff
@@ -223,7 +223,7 @@ static Value builtinArrayFrom(ExecutionState& state, Value thisValue, size_t arg
     if (C.isConstructor()) {
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: ditto

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 13]
```diff
@@ -268,7 +268,7 @@ static Value builtinArrayOf(ExecutionState& state, Value thisValue, size_t argc,
     Object* A;
     if (C.isConstructor()) {
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: ditto

File: src/runtime/GlobalObjectBuiltinWeakMap.cpp [hunk 14]
```diff
@@ -51,7 +51,7 @@ Value builtinWeakMapConstructor(ExecutionState& state, Value thisValue, size_t a
         adder = map->Object::get(state, ObjectPropertyName(state.context()->staticStrings().set)).value(state, map);
         // If IsCallable(adder) is false, throw a TypeError exception.
         if (!adder.isFunction()) {
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinWeakMap.cpp]: For readability / State expression explicitly
if (adder.isFunction() == false)

File: src/runtime/GlobalObjectBuiltinWeakSet.cpp [hunk 15]
```diff
@@ -53,7 +53,7 @@ Value builtinWeakSetConstructor(ExecutionState& state, Value thisValue, size_t a
         adder = set->Object::get(state, ObjectPropertyName(state.context()->staticStrings().add)).value(state, set);
         // If IsCallable(adder) is false, throw a TypeError exception.
         if (!adder.isFunction()) {
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinWeakSet.cpp]: ditto

File: src/runtime/Object.h [hunk 16]
```diff
@@ -587,6 +587,15 @@ class Object : public PointerValue {
 
     bool isConcatSpreadable(ExecutionState& state);
 
+    virtual Value call(ExecutionState& state, const Value& thisValue, const size_t argc, Value* argv)
```
Comment (yichoi) [inline src/runtime/Object.h]: Check pointer-type parameter / check before deferencing
ASSERT(argv != nullptr)

File: src/runtime/Object.h [hunk 17]
```diff
@@ -587,6 +587,15 @@ class Object : public PointerValue {
...
+    {
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+    virtual Object* construct(ExecutionState& state, const size_t argc, Value* argv, const Value& newTarget)
```
Comment (yichoi) [inline src/runtime/Object.h]: ditto

File: src/runtime/ProxyObject.cpp [hunk 18]
```diff
@@ -772,7 +772,7 @@ bool ProxyObject::set(ExecutionState& state, const ObjectPropertyName& propertyN
 }
 
 // https://www.ecma-international.org/ecma-262/6.0/#sec-proxy-object-internal-methods-and-internal-slots-call-thisargument-argumentslist
-Value ProxyObject::call(ExecutionState& state, const Value& callee, const Value& receiver, const size_t argc, Value* argv)
+Value ProxyObject::call(ExecutionState& state, const Value& receiver, const size_t argc, Value* argv)
```
Comment (yichoi) [inline src/runtime/ProxyObject.cpp]: ditto

File: src/runtime/ProxyObject.cpp [hunk 19]
```diff
@@ -812,9 +812,8 @@ Value ProxyObject::call(ExecutionState& state, const Value& callee, const Value&
     return FunctionObject::call(state, trap, handler, 3, arguments);
 }
 
-// FIXME newTarget is missing [[Construct]] ( argumentsList, newTarget)
 // https://www.ecma-international.org/ecma-262/6.0/#sec-proxy-object-internal-methods-and-internal-slots-construct-argumentslist-newtarget
-Object* ProxyObject::construct(ExecutionState& state, const size_t argc, Value* argv)
+Object* ProxyObject::construct(ExecutionState& state, const size_t argc, Value* argv, const Value& newTarget)
```
Comment (yichoi) [inline src/runtime/ProxyObject.cpp]: ditto

File: src/runtime/ProxyObject.cpp [hunk 20]
```diff
@@ -842,8 +841,7 @@ Object* ProxyObject::construct(ExecutionState& state, const size_t argc, Value*
     // b. Return Construct(target, argumentsList, newTarget).
     if (trap.isUndefined()) {
```
Comment (yichoi) [inline src/runtime/ProxyObject.cpp]: For readability / State expression explicitly
if (trap.isUndefined() == true)

File: src/runtime/ProxyObject.cpp [hunk 21]
```diff
@@ -842,8 +841,7 @@ Object* ProxyObject::construct(ExecutionState& state, const size_t argc, Value*
...
         ASSERT(target.isConstructor());
```
Comment (yichoi) [inline src/runtime/ProxyObject.cpp]: ditto

File: src/runtime/FunctionObject.cpp (hunk 2 above)
Comment (clover2123) [reply src/runtime/FunctionObject.cpp]: `argv` could be a null pointer, so I added `NULLABLE` keyword to notify it.

**Comment by clover2123** [comment]
@yichoi I fixed all your comments. Nullable `argv` is marked with `NULLABLE` keyword. Thank you for the review.

**Pull request merged**